src/sync/app.hpp

```diff
@@ -58,6 +55,181 @@ class App {
         realm::util::Optional<uint64_t> default_request_timeout_ms;
     };
 
+    /**
+     * A struct representing a user API key as returned by the App server.
+    */
+    struct UserAPIKey {
+        // The ID of the key.
+        ObjectId id;
+
+        // The actual key. Will only be included in
+        // the response when an API key is first created.
+        Optional<std::string> key;
+
+        // The name of the key.
+        std::string name;
+
+        // Whether or not the key is disabled.
+        bool disabled;
+    };
+
+    /**
+     * A client for the user API key authentication provider which
+     * can be used to create and modify user API keys. This
+     * client should only be used by an authenticated user.
+    */
+    class UserAPIKeyProviderClient {
+    public:
+        /**
+         * Creates a user API key that can be used to authenticate as the current user.
+         *
+         * - parameters:
+         *     - name: The name of the API key to be created.
+         *     - completion_block: A callback to be invoked once the call is complete.
+        */
+        void create_api_key(const std::string& name,
+                            std::function<void(UserAPIKey, Optional<AppError>)> completion_block);
+
+        /**
+         * Fetches a user API key associated with the current user.
+         *
+         * - parameters:
+         *     - id: The id of the API key to fetch.
+         *     - completion_block: A callback to be invoked once the call is complete.
+         */
+        void fetch_api_key(const ObjectId& id,
+                           std::function<void(UserAPIKey, Optional<AppError>)> completion_block);
+
+        /**
+         * Fetches the user API keys associated with the current user.
+         *
+         * - parameters:
+         *     - completion_block: A callback to be invoked once the call is complete.
+         */
+        void fetch_api_keys(std::function<void(UserAPIKey, Optional<AppError>)> completion_block);
+
+        /**
+         * Deletes a user API key associated with the current user.
+         *
+         * - parameters:
+         *     - id: The id of the API key to delete.
+         *     - completion_block: A callback to be invoked once the call is complete.
+         */
+        void delete_api_key(const ObjectId& id,
```

Comment:
It might be nice to pass in a populated `UserAPIKey` object for these functions instead of using an `ObjectId` directly; for one, it hides the implementation a bit, and also means that people can't pass in random ObjectIds.